src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+// Support for HC-SR04 style ultrasonic ranging modules. See
+// https://cdn.sparkfun.com/datasheets/Sensors/Proximity/HCSR04.pdf.
+
+use std::{
+    collections::HashMap,
+    num::NonZeroU32,
+    sync::{
+        atomic::{AtomicI64, Ordering},
+        Arc, Mutex,
+    },
+    time::Duration,
+};
+
+use crate::{
+    common::{
+        board::BoardType,
+        config::{AttributeError, ConfigType},
+        registry::{self, ComponentRegistry, Dependency},
+        sensor::{
+            GenericReadingsResult, Readings, Sensor, SensorResult, SensorT, SensorType,
+            TypedReadingsResult,
+        },
+        status::Status,
+    },
+    google, DoCommand,
+};
+
+use esp_idf_hal::{
+    delay::TickType,
+    gpio::{
+        enable_isr_service, init_isr_alloc_flags, AnyIOPin, Input, InterruptType, Output,
+        PinDriver, Pull,
+    },
+    task::notification::{Notification, Notifier},
+};
+
+use esp_idf_sys::{esp, gpio_isr_handler_add, gpio_isr_handler_remove};
+
+pub(crate) fn register_models(registry: &mut ComponentRegistry) {
+    if registry
+        .register_sensor("ultrasonic", &HCSR04Sensor::from_config)
+        .is_err()
+    {
+        log::error!("HCSR04Sensor is already registered");
+    }
+}
+
+struct IsrSharedState {
+    // The state machine used to track interrupts and compute the
+    // length of the echo pulse. It holds one of the following values:
+    //
+    // 0: Starting state, ready to take a reading
+    // i64 > 0: Millisecond timestamp of first edge of echo signal
+    // -1 : Lockout state, further interrupts will be ignored
+    timestamp: AtomicI64,
+
+    // The channel the ISR will use to communicate results back to waiters.
+    notifier: Arc<Notifier>,
+}
+
+#[derive(DoCommand)]
+pub struct HCSR04Sensor {
+    // We require a board in order to ensure that ISR setup has completed
+    _board: BoardType,
+
+    // The PinDriver to control the pin that triggers issuing a pulse.
+    trigger_pin: PinDriver<'static, AnyIOPin, Output>,
+
+    // The PinDriver used to listen for digital interrupts and measure
+    // the length of the echo pulse.
+    echo_interrupt_pin: PinDriver<'static, AnyIOPin, Input>,
```

Comment:
This does not seem to work if this `PinDriver` is  made `InputOutput`, so I can't use the `Esp32GPIOPin` stuff, since that unconditionally makes all pins `InputOutput`.